File: src/ros2_app_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"
#include "my_robot_interfaces/srv/reset_counter.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node // MODIFY NAME
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0)
    {
        number_subscriber_ = this->create_subscription<std_msgs::msg::Int64>(
            "number", 10, std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));

        reset_counter_service_ = this->create_service<my_robot_interfaces::srv::ResetCounter>(
            "reset_counter", std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
        
        RCLCPP_INFO(this->get_logger(), "number_counter has been started.");
    }

private:
    void callbackNumber(const std_msgs::msg::Int64::SharedPtr msg)
    {
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
    }

    void callbackResetCounter(const my_robot_interfaces::srv::ResetCounter::Request::SharedPtr request,
                              const my_robot_interfaces::srv::ResetCounter::Response::SharedPtr response)
    {
        if (request->reset_value >= 0)
        {
            counter_ = request->reset_value;
            RCLCPP_WARN(this->get_logger(), "Counter: %d", counter_);
            response->success = true;
        }
        else
        {
            RCLCPP_ERROR(this->get_logger(), "Requested value out of bounds!");
            response->success = false;
        }
    }

    int counter_;
    rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr number_subscriber_;
    rclcpp::Service<my_robot_interfaces::srv::ResetCounter>::SharedPtr reset_counter_service_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/ros2_app_py/ros2_app_py/number_counter.py
#!/usr/bin/env python3
from pickle import TRUE
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64
from my_robot_interfaces.srv import ResetCounter


class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")

        self.counter_ = 0

        self.sub_ = self.create_subscription(
            Int64, "number", self.callback_number, 10)

        self.reset_counter_service_ = self.create_service(
            ResetCounter, "reset_counter", self.callback_reset_counter) 

        self.get_logger().info("number_counter has been started.")

    def callback_number(self, msg):
        self.counter_ += msg.data
        self.get_logger().info("Counter: " + str(self.counter_))

    def callback_reset_counter(self, request, response):
        if request.reset_value >= 0:
            self.counter_ = request.reset_value
            self.get_logger().warn("Counter: " + str(self.counter_))
            response.success = True
        else:
            response.success = False
        return response



def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
